File: py/rover.py
from const import chars


class Rover:
    def __init__(self, name, x=0, y=0, direction="N", mars=object()):
        self.compass = ["N", "E", "S", "W"]
        self.mars = mars
        self.name = name
        self.direction = direction
        self.position = {"x": x, "y": y}
        self.travel_log = []
        self._create()

    def _logger(self):
        log = [coord for coord in self.position.values()]
        log.append(self.direction)

        return self.travel_log.append(tuple(log))

    def _create(self):
        name = self.name
        x = self.position["x"]
        y = self.position["y"]
        direction = self.direction
        grid = self.mars.grid if self.mars.grid else [[]]

        grid[y][x] = chars["rover"]

        print(f"• {name}: Landing position: {x}, {y} [{direction}]\n")

        return self._logger()

    def _output(self, kind, direction, moves):
        icons = {
            "turn": {
                "forth_arrow": "↻",
                "back_arrow": "↺",
                "forth_instruction": "right",
                "back_instruction": "left",
            },
            "move": {
                "forth_arrow": "↑",
                "back_arrow": "↓",
                "forth_instruction": "forwards",
                "back_instruction": "backwards",
            },
        }

        forth_arrow = icons[kind]["forth_arrow"]
        back_arrow = icons[kind]["back_arrow"]
        forth_instruction = icons[kind]["forth_instruction"]
        back_instruction = icons[kind]["back_instruction"]

        polarity = direction > 0
        arrow = forth_arrow if polarity else back_arrow
        instruction = forth_instruction if polarity else back_instruction

        x_moves = f" x{moves}" if moves > 1 else ""

        return (arrow, instruction, x_moves)

    def _turn(self, direction=1, moves=1):
        arrow, instruction, x_moves = self._output("turn", direction, moves)

        operator = "+" if direction else "-"
        now = self.compass.index(self.direction)
        total = eval(f"{now} {operator} {moves}")
        index = total % len(self.compass)

        self.direction = self.compass[index]

        print(f"{arrow} {self.name} turning {instruction}{x_moves} [{self.direction}]")

        return self.direction

    def _get_position(self, direction=1, moves=1):
        x = self.position["x"]
        y = self.position["y"]

        if direction:
            x += moves if self.direction == "E" else 0
            x -= moves if self.direction == "W" else 0
            y -= moves if self.direction == "N" else 0
            y += moves if self.direction == "S" else 0
        else:
            x -= moves if self.direction == "E" else 0
            x += moves if self.direction == "W" else 0
            y += moves if self.direction == "N" else 0
            y -= moves if self.direction == "S" else 0

        return (x, y)

    def _validate_move(self, direction=1):
        grid = self.mars.grid if self.mars.grid else [[]]

        x, y = self._get_position(direction, 1)

        if not 0 <= y < len(grid) or not 0 <= x < len(grid):
            print("\tError: Rover can't go over the edges.")
            return (x, y, False)

        if grid[y][x]:
            kind = "Obstacle" if grid[y][x] == chars["obstacle"] else "Rover"
            print(f"\t{self.name}: {kind} found [{x}, {y}]. Can't move.")
            return (x, y, False)

        return (x, y, True)

    def _move(self, direction=1, moves=1):
        arrow, instruction, x_moves = self._output("move", direction, moves)
        to_x, to_y = self._get_position(direction, moves)

        print(
            f'{arrow} {self.name} moving {instruction}{x_moves}...\
            \n\tFrom: [{self.position["x"]}, {self.position["y"]}] \
            \n\tTo:   [{to_x}, {to_y}]'
        )

        for n in range(moves):
            x, y, safe = self._validate_move(direction)

            if not safe:
                break

            self.mars.grid[y][x] = chars["rover"]
            self.mars.grid[self.position["y"]][self.position["x"]] = ""

            self.position["x"] = x
            self.position["y"] = y

            self._logger()

        return (self.position["x"], self.position["y"])

    def right(self, moves=1):
        return self._turn(moves=moves)

    def left(self, moves=1):
        return self._turn(0, moves)

    def forwards(self, moves=1):
        return self._move(moves=moves)

    def backwards(self, moves=1):
        return self._move(0, moves)

    def command(self, commands):
        valid = ["r", "l", "f", "b"]

        for command in commands:
            if command in valid:
                self.right() if command == valid[0] else 0
                self.left() if command == valid[1] else 0
                self.forwards() if command == valid[2] else 0
                self.backwards() if command == valid[3] else 0

    def done(self):
        name = self.name
        x = self.position["x"]
        y = self.position["y"]
        direction = self.direction
        log = self.travel_log

        print(
            f"\n• {name}: Final position: {x}, {y} [{direction}] \
            \n\t{name}: Travel Log:\n\t{log}\n"
        )

        return (x, y, direction)

File: js/node/rover.js
const {obsChar, roverChar} = require('./const');

module.exports = class Rover {
  constructor(
      name, x = 0, y = 0, direction = 'N', mars = {'grid': [[], []]},
  ) {
    this.compass = ['N', 'E', 'S', 'W'];
    this.direction = direction;
    this.mars = mars;
    this.name = name;
    this.position = {'x': x, 'y': y};
    this.travelLog = [];
    this.create();
  }

  logger() {
    const log = Object.values(this.position);
    log.push(this.direction);
    return this.travelLog.push(log);
  }

  create() {
    const direction = this.direction;
    const grid = this.mars.grid;
    const name = this.name;
    const x = this.position.x;
    const y = this.position.y;

    grid[y][x] = roverChar;

    console.log(`${name}: Landing position: ${x}, ${y} [${direction}] \n`);

    return this.logger();
  }

  output(type, direction, moves) {
    const icons = {
      'turn': {
        'forthArrow': '↻',
        'backArrow': '↺',
        'forthInstruction': 'right',
        'backInstruction': 'left',
      },
      'move': {
        'forthArrow': '↑',
        'backArrow': '↓',
        'forthInstruction': 'forwards',
        'backInstruction': 'backwards',
      },
    };

    const forthArrow = icons[type].forthArrow;
    const backArrow = icons[type].backArrow;
    const forthInstruction = icons[type].forthInstruction;
    const backInstruction = icons[type].backInstruction;

    const polarity = direction > 0;
    const arrow = polarity ? forthArrow : backArrow;
    const instruction = polarity ? forthInstruction : backInstruction;

    const xMoves = moves > 1 ? ` x${moves}` : '';

    return [arrow, instruction, xMoves];
  }

  turn(direction = 1, moves = 1) {
    const [arrow, instruction, xMoves] = this.output('turn', direction, moves);
    let [...compass] = this.compass;

    compass = !direction ? compass.reverse() : compass;

    const now = compass.indexOf(this.direction);
    const total = (now + moves) % compass.length;

    this.direction = compass[total];

    console.log(
        `${arrow} ${this.name} turning ${instruction}${xMoves}`,
        `[${this.direction}]`,
    );

    return this.direction;
  }

  getPosition(direction = 1, moves = 1) {
    let x = this.position.x;
    let y = this.position.y;

    if (direction) {
        this.direction === 'E' ? x += moves : 0;
        this.direction === 'W' ? x -= moves : 0;
        this.direction === 'N' ? y -= moves : 0;
        this.direction === 'S' ? y += moves : 0;
    } else {
        this.direction === 'E' ? x -= moves : 0;
        this.direction === 'W' ? x += moves : 0;
        this.direction === 'N' ? y += moves : 0;
        this.direction === 'S' ? y -= moves : 0;
    }

    return [x, y];
  }

  validateMove(direction = 1) {
    const grid = this.mars.grid;
    const [x, y] = this.getPosition(direction, 1);

    if (! (y >= 0 && y < grid.length) ||
        ! (x >= 0 && x < grid.length) ) {
      console.log('\tError: Rover can\'t go over the edges.');

      return [x, y, false];
    }

    if (grid[y][x]) {
      const type = grid[y][x] === obsChar ? 'Obstacle' : 'Rover';

      console.log(`\t${this.name}: ${type} found [${x}, ${y}]. Can't move.`);

      return [x, y, false];
    }

    return [x, y, true];
  }

  move(direction = 1, moves = 1) {
    const [arrow, instruction, xMoves] = this.output('move', direction, moves);
    const [toX, toY] = this.getPosition(direction, moves);

    console.log(
        `${arrow} ${this.name} moving ${instruction}${xMoves}... \
          \n\tFrom: [${this.position.x}, ${this.position.y}] \
          \n\tTo:   [${toX}, ${toY}]`);

    for (let i = 0; i < moves; i++) {
      const [x, y, safe] = this.validateMove(direction);

      if (!safe) break;

      this.mars.grid[y][x] = roverChar;
      this.mars.grid[this.position.y][this.position.x] = '';

      this.position.x = x;
      this.position.y = y;

      this.logger();
    }

    return [this.position.x, this.position.y];
  }

  right(moves = 1) {
    return this.turn(1, moves);
  };

  left(moves = 1) {
    return this.turn(0, moves);
  };

  forwards(moves = 1) {
    return this.move(1, moves);
  };

  backwards(moves = 1) {
    return this.move(0, moves);
  };

  command(commands) {
    const valid = ['r', 'l', 'f', 'b'];

    for (const command of commands) {
      if (valid.includes(command)) {
        command === valid[0] ? this.right() : null;
        command === valid[1] ? this.left() : null;
        command === valid[2] ? this.forwards() : null;
        command === valid[3] ? this.backwards() : null;
      }
    }
  };

  done() {
    const direction = this.direction;
    const name = this.name;
    const x = this.position.x;
    const y = this.position.y;

    console.log(
        `\n• ${name}: Final position: ${x}, ${y} [${direction}] \
        \n\t${name}: Travel Log:\n\t`,
        JSON.stringify(this.travelLog), '\n',
    );

    return [x, y, direction];
  }
};
